File: 2020/20_JurassicJigsaw/test_tile.lua
-- ======================================================================
-- Jurassic Jigsaw
--   Advent of Code 2020 Day 20 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ t i l e . l u a
-- ======================================================================
-- Test Tile for Advent of Code 2020 day 20, Jurassic Jigsaw

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local tile = require('tile')

-- ----------------------------------------------------------------------
--                                                              from_text
-- ----------------------------------------------------------------------

function from_text(text)
  -- Break the text into trimed, non-comment lines

  -- 1. We start with no lines
  local result = {}

  -- 2. Loop for lines in the text
  for line in text:gmatch('[^\r\n]+') do

    -- 3. But ignore blank and comment lines
    line = line:gsub("%s*$", "")
    if #line > 0 and "!" ~= line:sub(1, 1) then

      -- 4. Add the line
      table.insert(result, line)
    end
  end

  -- 5. Return a table of cleaned lines
  return result
end

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = [[
Tile 2311:
..##.#..#.
##..#.....
#...##..#.
####.#...#
##.##.###.
##...#.###
.#.#.#..##
..#....#..
###...#.#.
..###..###
]]

local EXAMPLE_DISPLAY = [[..##.#..#.
##..#.....
#...##..#.
####.#...#
##.##.###.
##...#.###
.#.#.#..##
..#....#..
###...#.#.
..###..###]]

local EXAMPLE_ROTATED = [[.#..#####.
.#.####.#.
###...#..#
#..#.##..#
#....#.##.
...##.##.#
.#...#....
#.#.##....
##.###.#.#
#..##.#...]]

local EXAMPLE_FLIP_HOR = [[#..##.#...
##.###.#.#
#.#.##....
.#...#....
...##.##.#
#....#.##.
#..#.##..#
###...#..#
.#.####.#.
.#..#####.]]

local BORDERS = "#..##.#...,.#..#.##..,.#..#####.,###..###.."

local EXAMPLE_IMAGE = [[#.###.#.
.#.##...
#...#...
..##.##.
....#.##
..#.##..
##...#..
#.####.#]]

local EXAMPLE_FLIP_VER = [[...#.##..#
#.#.###.##
....##.#.#
....#...#.
#.##.##...
.##.#....#
#..##.#..#
#..#...###
.#.####.#.
.#####..#.]]

local BORDERSV = "...#.##..#,###..###..,.#####..#.,.#..#.##.."

local IMAGE = {
  ".#.#..#.##...#.##..#####",
  "###....#.#....#..#......", 
  "##.##.###.#.#..######...",
  "###.#####...#.#####.#..#",
  "##.#....#.##.####...#.##",
  "...########.#....#####.#",
  "....#..#...##..#.#.###..",
  ".####...#..#.....#......",
  "#..#.##..#..###.#.##....",
  "#.####..#.####.#.#.###..",
  "###.#.#...#.######.#..##",
  "#.####....##..########.#",
  "##..##.#...#...#.#.#.#..",
  "...#..#..#.#.##..###.###",
  ".#.#....#.##.#...###.##.",
  "###.#...#..#.##.######..",
  ".#.#.###.##.##.#..#.##..",
  ".####.###.#...###.#..#.#",
  "..#.#..#..#.#.#.####.###",
  "#..####...#.#.#.###.###.",
  "#####..#####...###....##",
  "#.##..#..#...#..####...#",
  ".#.###..##..##..####.##.",
  "...###...##...#...#..###"
}

local MONSTER_IMAGE = [[.####...#####..#...###..
#####..#..#.#.####..#.#.
.#.#...#.###...#.##.##..
#.#.##.###.#.##.##.#####
..##.###.####..#.####.##
...#.#..##.##...#..#..##
#.##.#..#.#..#..##.#.#..
.###.##.....#...###.#...
#.####.#.#....##.#..#.#.
##...#..#....#..#...####
..#.##...###..#.#####..#
....#.##.#.#####....#...
..##.##.###.....#.##..#.
#...#...###..####....##.
.#.##...#.##.#.#.###...#
#.###.#..####...##..#...
#.###...#.##...#.######.
.###.###.#######..#####.
..##.#..#..#.#######.###
#.#..##.########..#..##.
#.#####..#.#...##..#....
#....##..#.#########..##
#...#.....#..##...###.##
#..###....##.#...##.##.#]]



-- ======================================================================
--                                                               TestTile
-- ======================================================================

function test_empty_init()
  -- Test the default Tile creation

  -- 1. Create default Tile object
  local myobj = tile:Tile()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(myobj.number, 0)
  luaunit.assertEquals(#myobj.rows, 0)
  luaunit.assertEquals(#myobj.flip, 0)
  luaunit.assertEquals(#myobj.alt, 0)

end

function test_text_init()
  -- Test the Tile object creation from text

  -- 1. Create Image object from text
  local myobj = tile:Tile({text=from_text(EXAMPLE_TEXT)})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 11)
  luaunit.assertEquals(myobj.number, 2311)
  luaunit.assertEquals(#myobj.rows, 10)
  luaunit.assertEquals(myobj.rows[1]:len(), 10)
  luaunit.assertEquals(myobj.rows[1], "..##.#..#.")
  luaunit.assertEquals(myobj.rows[10]:len(), 10)
  luaunit.assertEquals(myobj.rows[10], "..###..###")
  luaunit.assertEquals(#myobj.flip, 10)
  luaunit.assertEquals(table.concat(myobj.alt[BORDERS],'\n'), EXAMPLE_IMAGE)
  luaunit.assertEquals(myobj:count_alt(), 8)
    
  -- 3. Check methods
  luaunit.assertEquals(myobj:display(), EXAMPLE_DISPLAY)
  myobj:_initial_rotation()
  luaunit.assertEquals(myobj:display_rot(), EXAMPLE_DISPLAY)
  myobj:rotate()
  luaunit.assertEquals(myobj:display_rot(), EXAMPLE_ROTATED)
  myobj:_initial_flip()
  luaunit.assertEquals(myobj:display_flip(), EXAMPLE_ROTATED)
  myobj:flip_hor()
  luaunit.assertEquals(myobj:display_flip(), EXAMPLE_FLIP_HOR)
  local top, right, bottom, left, image = myobj:borders_and_image()
  luaunit.assertEquals(table.concat({top, right, bottom, left}, ','), BORDERS)
  luaunit.assertEquals(table.concat(image,'\n'), EXAMPLE_IMAGE)
  myobj:flip_ver()
  luaunit.assertEquals(myobj:display_flip(), EXAMPLE_FLIP_VER)
  top, right, bottom, left, image = myobj:borders_and_image()
  luaunit.assertEquals(table.concat({top, right, bottom, left}, ','), BORDERSV)
  luaunit.assertEquals(myobj:border(BORDERS, 'T'), "#..##.#...")
  luaunit.assertEquals(myobj:border(BORDERS, 'R'), ".#..#.##..")
  luaunit.assertEquals(myobj:border(BORDERS, 'B'), ".#..#####.")
  luaunit.assertEquals(myobj:border(BORDERS, 'L'), "###..###..")
end

function test_image()
  -- Test the default Tile creation

  -- 1. Create default Tile object
  local myobj = tile:Tile()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(myobj.number, 0)
  luaunit.assertEquals(#myobj.rows, 0)

  -- 3. Set the image
  myobj.rows = IMAGE
  
  -- 4. Methods
  luaunit.assertEquals(myobj:display(), table.concat(myobj.rows, '\n'))
  myobj:construct_alternatives(true)
  luaunit.assertEquals(#myobj.alt, 16)
  local found = false
  for _, altimg in ipairs(myobj.alt) do
    if table.concat(altimg,'\n') == MONSTER_IMAGE then
      found = true
      end
  end
  luaunit.assertEquals(found, true)
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                     t e s t _ t i l e . p y                    end
-- ======================================================================

File: 2023/16_TheFloorWillBeLava/test_tile.py

# ======================================================================
# The Floor Will Be Lava
#   Advent of Code 2023 Day 16 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                    t e s t _ t i l e . p y
# ======================================================================
"Test Tile for Advent of Code 2023 day 16, The Floor Will Be Lava"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import tile

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                               TestTile
# ======================================================================


class TestTile(unittest.TestCase):  # pylint: disable=R0904
    "Test Tile object"

    def test_default_init(self):
        "Test the default Tile creation"

        # 1. Create default Tile object
        myobj = tile.Tile()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.row, 0)
        self.assertEqual(myobj.col, 0)
        self.assertEqual(myobj.char, tile.EMPTY)
        self.assertEqual(myobj.activated, False)
        self.assertEqual(len(myobj.directions), 0)

    def test_empty_init(self):
        "Test the Tile object creation from text"

        # 1. Create Tile object from text
        myobj = tile.Tile(row=1, col=5, char=tile.EMPTY)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.row, 1)
        self.assertEqual(myobj.col, 5)
        self.assertEqual(myobj.char, tile.EMPTY)
        self.assertEqual(myobj.activated, False)
        self.assertEqual(len(myobj.directions), 0)

        # 3. Check methods
        self.assertEqual(myobj.next(tile.NORTH), (0, 5))
        self.assertEqual(myobj.next(tile.SOUTH), (2, 5))
        self.assertEqual(myobj.next(tile.EAST), (1, 6))
        self.assertEqual(myobj.next(tile.WEST), (1, 4))

        self.assertEqual(myobj.is_activated(), False)
        self.assertEqual(myobj.entered(tile.NORTH), [tile.NORTH])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.NORTH), [])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.SOUTH), [tile.SOUTH])
        self.assertEqual(myobj.entered(tile.EAST), [tile.EAST])
        self.assertEqual(myobj.entered(tile.WEST), [tile.WEST])
        self.assertEqual(len(myobj.directions), 4)

    def test_left_init(self):
        "Test the Tile object creation from text"

        # 1. Create Tile object from text
        myobj = tile.Tile(row=1, col=5, char=tile.LEFT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.row, 1)
        self.assertEqual(myobj.col, 5)
        self.assertEqual(myobj.char, tile.LEFT)
        self.assertEqual(myobj.activated, False)
        self.assertEqual(len(myobj.directions), 0)

        # 3. Check methods
        self.assertEqual(myobj.is_activated(), False)
        self.assertEqual(myobj.entered(tile.NORTH), [tile.EAST])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.NORTH), [])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.SOUTH), [tile.WEST])
        self.assertEqual(myobj.entered(tile.EAST), [tile.NORTH])
        self.assertEqual(myobj.entered(tile.WEST), [tile.SOUTH])
        self.assertEqual(len(myobj.directions), 4)

    def test_flat_init(self):
        "Test the Tile object creation from text"

        # 1. Create Tile object from text
        myobj = tile.Tile(row=1, col=5, char=tile.FLAT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.row, 1)
        self.assertEqual(myobj.col, 5)
        self.assertEqual(myobj.char, tile.FLAT)
        self.assertEqual(myobj.activated, False)
        self.assertEqual(len(myobj.directions), 0)

        # 3. Check methods
        self.assertEqual(myobj.is_activated(), False)
        self.assertEqual(myobj.entered(tile.NORTH), [tile.EAST, tile.WEST])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.NORTH), [])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.SOUTH), [tile.EAST, tile.WEST])
        self.assertEqual(myobj.entered(tile.EAST), [tile.EAST])
        self.assertEqual(myobj.entered(tile.WEST), [tile.WEST])
        self.assertEqual(len(myobj.directions), 4)

    def test_vert_init(self):
        "Test the Tile object creation from text"

        # 1. Create Tile object from text
        myobj = tile.Tile(row=1, col=5, char=tile.VERT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.row, 1)
        self.assertEqual(myobj.col, 5)
        self.assertEqual(myobj.char, tile.VERT)
        self.assertEqual(myobj.activated, False)
        self.assertEqual(len(myobj.directions), 0)

        # 3. Check methods
        self.assertEqual(myobj.is_activated(), False)
        self.assertEqual(myobj.entered(tile.NORTH), [tile.NORTH])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.NORTH), [])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.SOUTH), [tile.SOUTH])
        self.assertEqual(myobj.entered(tile.EAST), [tile.NORTH, tile.SOUTH])
        self.assertEqual(myobj.entered(tile.WEST), [tile.NORTH, tile.SOUTH])
        self.assertEqual(len(myobj.directions), 4)

    def test_right_init(self):
        "Test the Tile object creation from text"

        # 1. Create Tile object from text
        myobj = tile.Tile(row=1, col=5, char=tile.RIGHT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.row, 1)
        self.assertEqual(myobj.col, 5)
        self.assertEqual(myobj.char, tile.RIGHT)
        self.assertEqual(myobj.activated, False)
        self.assertEqual(len(myobj.directions), 0)

        # 3. Check methods
        self.assertEqual(myobj.is_activated(), False)
        self.assertEqual(myobj.entered(tile.NORTH), [tile.WEST])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.NORTH), [])
        self.assertEqual(myobj.is_activated(), True)
        self.assertEqual(len(myobj.directions), 1)
        self.assertEqual(myobj.entered(tile.SOUTH), [tile.EAST])
        self.assertEqual(myobj.entered(tile.EAST), [tile.SOUTH])
        self.assertEqual(myobj.entered(tile.WEST), [tile.NORTH])
        self.assertEqual(len(myobj.directions), 4)

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                     t e s t _ t i l e . p y                    end
# ======================================================================
